### ISSUE TITLE
Error installing addon, "ShouldLog" error

### ISSUE THREAD

**Issue by meandavejustice** [mozilla/testpilot#1070]
- Linux 4.2.0
- Nightly

Addon is throwing an error on install.

This is after uninstalling TxP addon, navigating to https://testpilot.firefox.com/ clicking the "Install the Test Pilot Add On" button

error msg and stack trace here => https://gist.github.com/meandavejustice/db804445d6ec12988a5f509db7f584c9

I've tracked it down to this line https://github.com/mozilla/testpilot/blob/master/addon/lib/webextension-channels.js#L130

when above call to `console.error` is removed, the addon installs as intended.

**Comment by jaredhirsch** [comment]
I think that console.error line isn't the root cause. It will only be reached if `notifyPing` throws.

Uninstalling the TxP addon will unload all of the code. If we had a pending `notifyPing` that wasn't canceled, and it failed for some reason (maybe it timed out), then the code might try to call `console.error`, but fail, because Console.jsm has been unloaded behind the scenes (or maybe it wasn't loaded to begin with? not sure if SDK provides `console` for free).

The traceback looks like a Promise threw in response to a NetUtil error (see the very bottom of the trace), so that's my guess: we need to clean up in-flight network requests before uninstalling.

I'll try to repro now by uninstalling, then reinstalling.

**Comment by jaredhirsch** [comment]
TLDR: I think there are two bugs here, but I don't think they would affect ordinary users. The bugs would probably only affect someone who uninstalled, then immediately tried to reinstall, the Test Pilot add-on: not a common use case.

The second traceback below points to us trying to use the pageMod after its connection is closed. It's possible that pageMod throws the error, then fails to detach. If that's the case, then Dave's symptoms (uninstall, reinstall, but get stuck at the 'please install the add-on' page) would make sense: the pageMod can't communicate with the page because there's another one clogging the connection.

It also looks like there may be an error with disconnecting the webextension-channel listener from the BroadcastChannel, this is suggested by the first traceback.

---

When I go through these steps:
- go to testpilot.firefox.com
- sign into FxA
- install test pilot add-on through testpilot.ff.com website
- **go to about:addons and uninstall test pilot add-on**
- go back to testpilot.firefox.com
- reinstall test pilot add-on through testpilot.ff.com website

I see this error 7 times in the browser console:

```
TypeError: Property 'handleEvent' is not callable. (unknown)
```

I think the `(unknown)` means that the code is getting called, but it's been unloaded (so there is no corresponding file that devtools can locate). Probably dead code that wasn't properly detached and is not GC-able.

It looks like the only occurrence of `handleEvent` in the add-on code is in `lib/webextension-channels.js`. My guess is that the handlers are not getting detached from the BroadcastChannel on uninstall, so that when you reinstall, the old code is getting called along with / instead of the new code. (I didn't observe this error when I disabled and re-enabled the add-on. It seems specific to uninstalling and reinstalling.)

---

In contrast, when I do the same basic thing (install, then uninstall), but I uninstall the add-on through the website, I see a different weird error that seems related to incomplete network cleanup on uninstall:
- go to testpilot.firefox.com
- sign into FxA
- install test pilot add-on through testpilot.ff.com website
- click 'choose your features'
- **go to TxP website menu and choose 'uninstall test pilot'**
- confirm uninstall in the modal on the site

I consistently see this error in the browser console:

```
testpilot-addon:TypeError: workers.get(...) is undefined
Stack trace:
workerFor@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/page-mod.js:29:26
WorkerHost/</accessorProp.get@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/content/utils.js:72:33
Router.prototype.send@resource://gre/modules/commonjs/toolkit/loader.js -> resource://testpilot-addon/index.js:318:3
exports.onUnload@resource://gre/modules/commonjs/toolkit/loader.js -> resource://testpilot-addon/index.js:598:14
unload/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/system/unload.js:58:7
unload@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/system/unload.js:56:3
onunload@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/system/unload.js:77:5
Observer<.observe@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/system/events.js:81:7
unload@resource://gre/modules/commonjs/toolkit/loader.js:795:3
unload/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/addon/bootstrap.js:153:9
unload@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/addon/bootstrap.js:149:12
shutdown@resource://gre/modules/commonjs/toolkit/loader.js -> resource://gre/modules/commonjs/sdk/addon/bootstrap.js:146:12
this.XPIProvider.callBootstrapMethod@resource://gre/modules/addons/XPIProvider.jsm:4675:9
this.XPIProvider.uninstallAddon@resource://gre/modules/addons/XPIProvider.jsm:4964:11
AddonWrapper.prototype.uninstall@resource://gre/modules/addons/XPIProvider.jsm:7198:5
uninstallSelf/<@resource://gre/modules/commonjs/toolkit/loader.js -> resource://testpilot-addon/index.js:394:43
safeCall@resource://gre/modules/AddonManager.jsm:179:5
makeSafe/<@resource://gre/modules/AddonManager.jsm:195:25
Handler.prototype.process@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:937:23
this.PromiseWalker.walkerLoop@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:816:7
this.PromiseWalker.scheduleWalkerLoop/<@resource://gre/modules/Promise.jsm -> resource://gre/modules/Promise-backend.js:750:11
```

I think what's happening here is:
- our add-on is getting uninstalled
- `onUnload` gets called in index.js
- `onUnload` calls `if (app) app.send('addon-self:uninstalled')` but the pageMod is already disconnected
- pageMod tries to send the message, fails, and doesn't clean up after itself. maybe it stays attached? not sure.

**Comment by jaredhirsch** [comment]
I'd guess the bug in the description is related to the `handleEvent` bug, because the traceback in the description points at a `console.error` line inside `notifyPing`, which is called by `handleEvent`.

I'll look a bit further and break this issue into separate bugs tomorrow.

**Comment by ghost** [comment]
thanks for the investigation, @6a68!

**Comment by jaredhirsch** [comment]
@SoftVision-PaulOiegas could you try to reproduce this bug?

**Comment by SoftVision-CosminMuntean** [comment]
We have tested this issue on the latest Nightly build and latest Firefox release on Windwos 7 x64 and Ubuntu 14.04 x64, but we could not reproduce it.

The following scenarios were used for testing:

[Prerequisites]:
- Use a clean profile without Test Pilot add-on installed.

Scenario 1:
1. Go to https://testpilot.firefox.com/ .
2  Click on the big green "Install the Test Pilot Add-on" button.
3. After the Test Pilot add-on is installed go to "about:addons".
4. Remove Test Pilot add-on.
5. Go back to https://testpilot.firefox.com/ .
6. Reinstall the Test Pilot add-on.

This are the errors displayed in the browser console:
- Windows 7 x64:

![browser console 1](https://cloud.githubusercontent.com/assets/19226534/17484132/505cdc40-5d91-11e6-9a7d-c17edbb56612.png)

<hr>

Scenario 2:
1. Go to https://testpilot.firefox.com/ .
2  Click on the big green "Install the Test Pilot Add-on" button.
3. After the Test Pilot add-on is installed go to https://testpilot.firefox.com/experiments .
4. Click on the "Settings" button.
5. From the displayed context menu click "Uninstall Test Pilot" option.
6. A pop-up window is displayed, and click the "Proceed" button.
7. After the add-on is uninstalled go back to https://testpilot.firefox.com/ .
8. Reinstall Test Pilot add-on.

This are the errors displayed in the browser console:
- Windows 7 x64:

![browser console 2](https://cloud.githubusercontent.com/assets/19226534/17484137/53ddce74-5d91-11e6-8a0f-a406c0c8762b.png)

<hr>

We have also tested this issue using a Nightly version with 32 bit and we encounter the same errors.

**Comment by ghost** [comment]
Closing since we haven't seen it for a long time
